File: scala/day22_part1_2024.scala

import scala.io.Source

def nextSecret(s: Long): Long = {
  var x = s * 64
  var result = s ^ x
  result &= 0xFFFFFF
  x = result / 32
  result ^= x
  result &= 0xFFFFFF
  x = result * 2048
  result ^= x
  result &= 0xFFFFFF
  result
}

@main def main(): Unit = {
  val buyers = Source.fromFile("input.txt").getLines().filter(_.nonEmpty).map(_.toLong).toList
  val total = buyers.map { b =>
    var s = b
    for (_ <- 0 until 2000) {
      s = nextSecret(s)
    }
    s
  }.sum
  println(total)
}

File: ruby/day22_part1_2024.rb

def next_secret(s)
  x = s * 64
  s ^= x
  s &= 0xFFFFFF
  x = s / 32
  s ^= x
  s &= 0xFFFFFF
  x = s * 2048
  s ^= x
  s &= 0xFFFFFF
  s
end

total = File.readlines("input.txt").compact.map(&:to_i).sum do |buyer|
  s = buyer
  2000.times { s = next_secret(s) }
  s
end

puts total
